### PR TITLE
Add codegen for view model factories

### PR DESCRIPTION
**Pull request by aitorvs**
Task/Issue URL: https://app.asana.com/0/414730916066338/1202075724102108/f

### Description
This PR adds an Anvil kotlin compiler that generates `ViewModel` factories.

Before:

```kotlin
class MyViewModel constructor(...) {}

@ContributesMultibinding(AppScope::class)
class MyViewModelFactory @Inject constructor(...) : ViewModelFactoryPlugin {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T? {
        with(modelClass) {
            return when {
                isAssignableFrom(MyViewModelFactory::class.java) → {...}
                else -> null
            }
        }
    }
}
```

After:
```kotlin
@ContributesViewModel(AppScope::class)
class MyViewModel @Inject constructor(...) {}
```

### Steps to test this PR

_Test ViewModel signature validations_
- [x] Go to any ViewModel and remove the `@Inject` annotation from the constructor
- [x] Try to build and verify an error is thrown saying that the view model class `...must have an @Inject constructor`
- [x] Re-add the `@Inject` annotation
- [x] Go to `BrowserTabViewModel` and in line `159` add the default dispatcher, ie. `private val dispatchers: DispatcherProvider = DefaultDispatcherProvider()`,
- [x] Try to build and verify an error is thrown saying that the view model class `...constructor parameters must not have default values`
- [x] revert the change
- [x] go to any ViewModel and annotate it with `@SingleInstanceIn(...)`
- [x] Try to build and verify an error is thrown saying that the view model class `...cannot be annotated with @SingleInstanceIn`

_Smoke tests_
- [x] Install form this branch
- [x] Perform smoke tests for both browser and AppTP

### CODE REVIEW

**Comment by aitorvs** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #1850** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/1850" target="_blank">Graphite</a>  (this PR)

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/1850?utm_source=stack-comment).

**Comment by aitorvs** [comment]
FYI: The test that is failing in CI is the flaky `whenGetDatabasePathOnDeviceThenPathNotEmpty`

File: vpn/src/main/java/dummy/ui/VpnControllerViewModel.kt [hunk 1]
```diff
@@ -20,16 +20,20 @@ import android.content.Context
...
+import com.duckduckgo.di.scopes.AppScope
 import com.duckduckgo.mobile.android.vpn.model.VpnState
 import com.duckduckgo.mobile.android.vpn.model.VpnTracker
 import com.duckduckgo.mobile.android.vpn.service.TrackerBlockingVpnService
 import com.duckduckgo.mobile.android.vpn.stats.AppTrackerBlockingStatsRepository
 import com.duckduckgo.mobile.android.vpn.stats.AppTrackerBlockingStatsRepository.DataStats
 import kotlinx.coroutines.flow.map
+import javax.inject.Inject
 
-class VpnControllerViewModel(
```
Comment (karlenDimla) [inline vpn/src/main/java/dummy/ui/VpnControllerViewModel.kt:32]: Do we still need the VpnViewModelFactory? I think it wasn't removed yet.

**Review by karlenDimla — APPROVED** [review]
All changes make sense and great job removing a lot of code!
I smoke tested known pages in the app/widget/AppTP and everything works as expected.

**Pull request merged**